pkg/fqdn/messagehandler/message_handler.go

```diff
@@ -311,7 +241,106 @@ func (b *dnsRequestHandler) NotifyOnDNSMsg(
+	//                                       |No selectors need updating|
+	// T3 ----> wg := UpdatePolicyMaps()     +--------------------------+
+	//
+	// T4 ----> mutex.Unlock()               mutex.Lock() / mutex.Unlock() <-- T4
+	//
+	// T5 ----> wg.Wait()                    DNS released back to pod    <-- T5
+	//                                                    |
+	// T6 --> DNS released back to pod                    |
+	//              |                                     |
+	//              |                                     |
+	//              v                                     v
+	//       Traffic flows fine                   Leads to policy drop until T6
+	//
+	// Note how G2 releases the DNS msg back to the pod at T5 because
+	// UpdateGenerateDNS() was a no-op. It's a no-op because G1 had executed
+	// UpdateGenerateDNS() first at T1 and performed the necessary policy
+	// updates for the response IPs. Due to G1 performing all the work
+	// first, G2 executes T4 also as a no-op and releases the msg back to the
+	// pod at T5 before G1 would at T6.
+	//
+	// We do not do a `defer unlock()` here, as we should release the lock before
+	// doing final bookkeeping.
+	mutexAcquireStart := time.Now()
+	h.nameManager.LockName(qname)
+
+	if d := time.Since(mutexAcquireStart); d >= option.Config.DNSProxyLockTimeout {
+		h.logger.Warn(fmt.Sprintf("Name lock acquisition time took longer than expected. Potentially too many parallel DNS requests being processed, consider adjusting --%s and/or --%s", option.DNSProxyLockCount, option.DNSProxyLockTimeout),
+			logfields.DNSName, qname,
+			logfields.Duration, d,
+			logfields.Expected, option.Config.DNSProxyLockTimeout,
+		)
+	}
+
+	h.logger.Debug("Recording DNS lookup in endpoint specific cache", logfields.EndpointID, ep.ID)
+
+	// This must happen before the NameManager update below, to ensure that
+	// this data is included in the serialized Endpoint object.
+	// We also need to add to the cache before we purge any matching zombies
+	// because they are locked separately and we want to keep the allowed IPs
+	// consistent if a regeneration happens between the two steps. If an update
+	// doesn't happen in the case, we play it safe and don't purge the zombie
+	// in case of races.
+	if updated := ep.DNSHistory.Update(lookupTime, qname, responseIPs, int(TTL)); updated {
+		ep.DNSZombies.ForceExpireByNameIP(lookupTime, qname, responseIPs...)
+		ep.SyncEndpointHeaderFile()
+	}
+
+	h.logger.Debug("Updating DNS name in cache from response to query",
+		logfields.DNSName, qname,
+		logfields.IPAddrs, responseIPs,
+	)
+
+	updateCtx, updateCancel := context.WithTimeout(h.ctx, option.Config.FQDNProxyResponseMaxDelay)
+	defer updateCancel()
+	updateStart := time.Now()
+
+	dpUpdates := h.nameManager.UpdateGenerateDNS(updateCtx, lookupTime, qname, &fqdn.DNSIPRecords{
+		IPs: responseIPs,
+		TTL: int(TTL),
+	})
+
+	stat.PolicyGenerationTime.End(true)
+	stat.DataplaneTime.Start()
```

Comment:
nit (optional): For better readability, I would consider moving the `defer stat.DataplaneTime.End(true)` here